src/shapes/shape.rs

```diff
@@ -65,19 +65,23 @@ pub trait Dtype:
     + std::ops::MulAssign
     + std::ops::DivAssign
     + num_traits::FromPrimitive
+    + num_traits::Float
+    + rand_distr::uniform::SampleUniform
 {
 }
 impl Dtype for f32 {}
 impl Dtype for f64 {}
-impl Dtype for usize {}
 
 /// Represents something that has a [Dtype].
 pub trait HasDtype {
     type Dtype: Dtype;
 }
 
 /// Represents a single dimension of a multi dimensional [Shape]
-pub trait Dim: 'static + Copy + Clone + std::fmt::Debug + Send + Sync + Eq + PartialEq {
+pub trait Dim:
+    'static + Copy + Clone + std::fmt::Debug + Send + Sync + Eq + PartialEq + Default
```

Comment:
I had left Default off of both Dim & Shape purposely - mainly for runtime dimensions. While nothing in dfdx depends on Default impl for shapes anymore, I could see someone using this in a generic context, which would be misleading.

I think you can probably revert these changes to shape, and instead just use ConstShape for TensorOptions like I mention above. This does limit all the tensors to be compile time known for nn, but that's what they are for the foreseeable future anyway.

If we change nn to support runtime dimensions we can address this there. I suspect it'll require a lot of api changes anyway